src/BufferCopy/ColorBufferToRDRAM.cpp

```diff
@@ -234,17 +234,53 @@ bool ColorBufferToRDRAM::_prepareCopy(u32& _startAddress)
 	return true;
 }
 
-u8 ColorBufferToRDRAM::_RGBAtoR8(u8 _c) {
+// Precalculated 4x4 bayer matrix values for 5Bit
+const s32 thresholdMapBayer [4][4] = {
+	{-4,  2, -3,  4 },
+	{ 0, -2,  2, -1 },
+	{-3,  3, -4,  3 },
+	{ 1, -1,  1, -2 }
+};
+
+// Precalculated 4x4 magic square matrix values for 5Bit
+const s32 thresholdMapMagicSquare [4][4] = {
+	{-4,  2,  2, -1 },
+	{ 3, -2, -3,  1 },
+	{-3,  0,  4, -2 },
+	{ 3, -1, -4,  1 }
+};
+
+u8 ColorBufferToRDRAM::_RGBAtoR8(u8 _c, u32 x, u32 y) {
 	return _c;
 }
 
-u16 ColorBufferToRDRAM::_RGBAtoRGBA16(u32 _c) {
-	RGBA c;
+u16 ColorBufferToRDRAM::_RGBAtoRGBA16(u32 _c, u32 x, u32 y) {
+	union RGBA c;
 	c.raw = _c;
-	return ((c.r >> 3) << 11) | ((c.g >> 3) << 6) | ((c.b >> 3) << 1) | (c.a == 0 ? 0 : 1);
+	
+	if(gDP.otherMode.colorDither <= 1 && 
+		((config.frameBufferEmulation.nativeResFactor != 1 	&& config.generalEmulation.ditheringMode >= 3) 
+		|| config.generalEmulation.ditheringMode < 3)) {
+		s32 threshold = 0;
```

Comment:
I don't understand this condition.
gDP.otherMode.colorDither <= 1 is clear, the rest is not.

This condition needed to disable double dithering, that is dithering already dithered image.
Image can be already dithered only when native res is enabled and config.generalEmulation.ditheringMode >= 3
because dithering disappears when hi-res image scaled down. So, the condition is
```
gDP.otherMode.colorDither <= 1 // ordered grid dithering enabled in othermode
&& !(config.frameBufferEmulation.nativeResFactor == 1 && config.generalEmulation.ditheringMode >= 3) // image is NOT already dithered

```
or
```
gDP.otherMode.colorDither <= 1 // ordered grid dithering enabled in othermode
&& (config.frameBufferEmulation.nativeResFactor != 1 || config.generalEmulation.ditheringMode < 3)

```